changes/2020-07-06_clarify-streaming-encrypt-decrypt/change.md

```diff
@@ -0,0 +1,154 @@
+Specifically, we attempt to be completely language agnostic and frame streaming as a way for
+bytes to move sequentially, over time.
+
+With this framing, the things the encrypt and decrypt operations care about are:
+
+- Is the caller allowed to send any more data?
+- What data is currently available to me?
+
+In the context of the encrypt or decrypt operation, we specify the conditions
+for when the operation MAY or MUST perform some processing.
+These conditions include how many input bytes are avaiable,
+whether more bytes MAY be available,
+at what step in the operation you are in,
+and the result of various calculations.
+
+Additionally, we use similar conditions to specify when the output of the operation MAY be released.
+
+This change SHOULD NOT introduce any language which is implementation specific for
+a specific streaming interface.
+Any streaming interface which is capable of moving bytes over time from one place to another,
+and is able to indicate EOF MUST be able to implement this specification as written.
+
+### How should verification and signing be performed?
+
+In order to not require holding the entire message in memory to either create
+or verify a signature, this change specifies that streaming use cases input the message
+to the signature algorithm in parts, as it is being parsed or serialized.
+
+As such, this change states that once the header or a frame is serialized or parsed,
+it MUST be inputted to the signature algorithm (if included) such that it
+isn't required to remain in memory for the duration of the operation.
+
+### How and when should plaintext be released?
+
+For one-shot cases,
+where the entire message is available in memory,
+it is obvious that the operation MUST NOT release any plaintext until the
+signature (if included) is verified.
+
+For the streaming case, this gets more complicated.
+Streaming is a way to operate on large messages without requiring the entire message to be
+loaded into memory.
+This allows the processing of arbitrarily long messages with finite memory.
+However, when processing messages with a signature,
+the operation obtains plaintext in parts as it processes the message,
+but has no way of knowing whether the message is verified until the end of the operation.
+So what should we do with the unverified plaintext decrypted thus far?
+Because streaming is being used, we MUST NOT buffer the entire plaintext,
+as that would defeat the point of streaming.
+Thus, the plaintext needs to go somewhere other than memory.
+Because this is a client-side library, we MUST give it to the customer.
+
+In the case of any decryption failure,
+the operation MUST zero out any avilable but unreleaed plaintext.
+The customer MUST understand that all released plaintext is unverified until the operation
+completes and succeeds.
+The customer MUST discard any released plaintext if the operation fails, and MUST roll back
+any processing done with that released plaintext.
```

Comment:
Is there a way we can enforce these statements?

Reply:
There isn't. This is a contract that the customer must understand for stream-decrypting messages with a message signature.